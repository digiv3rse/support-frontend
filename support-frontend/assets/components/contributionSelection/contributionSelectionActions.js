// @flow

// ----- Imports ----- //

import type { ContributionType } from 'helpers/contributions';
import type { CountryGroupId } from 'helpers/internationalisation/countryGroup';


// ----- Types ----- //

export type Action =
  | { type: 'SET_CONTRIBUTION_TYPE', scope: string, contributionType: ContributionType, countryGroupId: CountryGroupId }
  | { type: 'SET_AMOUNT', scope: string, amount: string }
  | { type: 'SET_AMOUNT_FOR_CONTRIBUTION_TYPE', scope: string, contributionType: ContributionType, amount: string }
  | { type: 'SET_CUSTOM_AMOUNT', scope: string, amount: string, countryGroupId: CountryGroupId };


// ----- Action Creators ----- //

function contributionSelectionActionsFor(scope: string): Object {

  return {
    setContributionType(
      contributionType: ContributionType,
      countryGroupId: CountryGroupId,
    ): Action {
      return {
        type: 'SET_CONTRIBUTION_TYPE', scope, contributionType, countryGroupId,
      };
    },
    setAmount(amount: string): Action {
      return { type: 'SET_AMOUNT', scope, amount };
    },
    setAmountForContributionType(contributionType: ContributionType, amount: string): Action {
      return {
        type: 'SET_AMOUNT_FOR_CONTRIBUTION_TYPE',
        scope,
        contributionType,
        amount,
      };
    },
    setCustomAmount(amount: string, countryGroupId: CountryGroupId): Action {
      return {
        type: 'SET_CUSTOM_AMOUNT', scope, amount, countryGroupId,
      };
    },
  };

}


// ----- Exports ----- //

export { contributionSelectionActionsFor };
