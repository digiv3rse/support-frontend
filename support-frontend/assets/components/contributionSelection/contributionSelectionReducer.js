// @flow

// ----- Imports ----- //

import {
  parseContribution,
  validateContribution,
  config,
} from 'helpers/contributions';
import type { CountryGroupId } from 'helpers/internationalisation/countryGroup';

import type {
  ContributionType,
  ContributionError,
  ValidationError,
} from 'helpers/contributions';

import type { Action } from './contributionSelectionActions';


// ----- Types ----- //

export type State = {
  contributionType: ContributionType,
  oneOffAmount: string,
  monthlyAmount: string,
  annualAmount: string,
  isCustomAmount: boolean,
  customAmount: ?number,
  error: ?ContributionError,
};


// ----- Functions ----- //

// Returns a countryGroupId-specific initial state for the reducer.
function getInitialState(countryGroupId: CountryGroupId): State {
  return {
    contributionType: 'MONTHLY',
    oneOffAmount: config[countryGroupId].ONE_OFF.default.toString(),
    monthlyAmount: config[countryGroupId].MONTHLY.default.toString(),
    annualAmount: config[countryGroupId].ANNUAL.default.toString(),
    customAmount: null,
    isCustomAmount: false,
    error: null,
  };
}

// Changes the amount of the currently selected type of contribution.
function updatePredefinedAmount(state, contributionType: ContributionType, newAmount: string): State {

  const resetCustom = { isCustomAmount: false, error: null };

  switch (contributionType) {
    case 'ONE_OFF':
      return { ...state, ...resetCustom, oneOffAmount: newAmount };
    case 'MONTHLY':
      return { ...state, ...resetCustom, monthlyAmount: newAmount };
    case 'ANNUAL':
      return { ...state, ...resetCustom, annualAmount: newAmount };
    default:
      return state;
  }

}

// Re-parses the custom amount when the contribution type is changed.
function checkCustomAmount(
  isCustomAmount: boolean,
  customAmount: ?number,
  contributionType: ContributionType,
  countryGroupId: CountryGroupId,
): ?{ error: ?ValidationError } {

  if (isCustomAmount && customAmount) {

    const error = validateContribution(customAmount, contributionType, countryGroupId);
    return { error };

  }

  return null;

}

function parseCustomAmount(
  amount: string,
  contributionType: ContributionType,
  countryGroupId: CountryGroupId,
): { customAmount: ?number, error: ?ContributionError } {

  const parsed = parseContribution(amount);

  if (!parsed.valid) {
    return { customAmount: null, error: parsed.error };
  }

  return {
    customAmount: parsed.amount,
    error: validateContribution(parsed.amount, contributionType, countryGroupId),
  };

}


// ----- Selectors ----- //

// Retrieves the amount for the currently chosen contribution type.
function getAmount(state: State): string {

  if (state.isCustomAmount && state.customAmount) {
    return state.customAmount.toString();
  }

  switch (state.contributionType) {
    case 'ONE_OFF':
      return state.oneOffAmount;
    case 'MONTHLY':
      return state.monthlyAmount;
    default:
    case 'ANNUAL':
      return state.annualAmount;
  }

}


// ----- Reducer ----- //

function contributionSelectionReducerFor(scope: string, countryGroupId: CountryGroupId): Function {

  const initialState = getInitialState(countryGroupId);

  function contributionSelectionReducer(state: State = initialState, action: Action): State {

    if (action.scope !== scope) {
      return state;
    }

    switch (action.type) {
      case 'SET_CONTRIBUTION_TYPE':

        return {
          ...state,
          contributionType: action.contributionType,
          ...checkCustomAmount(
            state.isCustomAmount,
            state.customAmount,
            action.contributionType,
            action.countryGroupId,
          ),
        };

      case 'SET_AMOUNT':
        return updatePredefinedAmount(state, state.contributionType, action.amount);

      case 'SET_AMOUNT_FOR_CONTRIBUTION_TYPE':
        return updatePredefinedAmount(state, action.contributionType, action.amount);

      case 'SET_CUSTOM_AMOUNT':

        return {
          ...state,
          isCustomAmount: true,
          ...parseCustomAmount(action.amount, state.contributionType, action.countryGroupId),
        };

      default:
        return state;
    }

  }

  return contributionSelectionReducer;

}


// ----- Exports ----- //

export {
  contributionSelectionReducerFor,
  getAmount,
};
