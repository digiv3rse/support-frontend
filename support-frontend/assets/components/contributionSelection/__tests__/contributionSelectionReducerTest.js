// @flow

// ----- Imports ----- //

import { contributionSelectionReducerFor } from '../contributionSelectionReducer';
import { contributionSelectionActionsFor } from '../contributionSelectionActions';


// ----- Tests ----- //

describe('Contributions Selection reducer', () => {

  const PREFIX = 'MY_PREFIX';
  const reducer = contributionSelectionReducerFor(PREFIX, 'GBPCountries');
  const usReducer = contributionSelectionReducerFor(PREFIX, 'UnitedStates');
  const actions = contributionSelectionActionsFor(PREFIX);

  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('should return the initial state for another country group', () => {
    expect(usReducer(undefined, {})).toMatchSnapshot();
  });

  it('should not change the state for an action with the wrong scope', () => {

    const otherActions = contributionSelectionActionsFor('OTHER_PREFIX');

    const changeContributionType = otherActions.setContributionType('ONE_OFF');
    expect(reducer(undefined, changeContributionType)).toMatchSnapshot();

    const changeAmount = otherActions.setAmount('42');
    expect(reducer(undefined, changeAmount)).toMatchSnapshot();

    const changeCustomAmount = otherActions.setCustomAmount('42');
    expect(reducer(undefined, changeCustomAmount)).toMatchSnapshot();

  });

  it('should handle SET_CONTRIBUTION_TYPE', () => {

    const contributionType = 'ONE_OFF';

    const changeContributionType = actions.setContributionType(contributionType);
    expect(reducer(undefined, changeContributionType).contributionType).toEqual(contributionType);

  });

  it('should re-parse custom amounts when SET_CONTRIBUTION_TYPE is fired for MONTHLY', () => {

    const contributionType = 'MONTHLY';
    const initialState = {
      contributionType: 'ONE_OFF',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: '1',
      isCustomAmount: true,
      error: null,
    };

    const changeContributionType = actions.setContributionType(contributionType, 'GBPCountries');
    const newState = reducer(initialState, changeContributionType);
    expect(newState.error).toEqual('TooLittle');

  });

  it('should re-parse custom amounts when SET_CONTRIBUTION_TYPE is fired for ONE_OFF', () => {

    const contributionType = 'ONE_OFF';
    const initialState = {
      contributionType: 'MONTHLY',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: '1',
      isCustomAmount: true,
      error: 'TooLittle',
    };

    const changeContributionType = actions.setContributionType(contributionType, 'GBPCountries');
    const newState = reducer(initialState, changeContributionType);
    expect(newState.error).toBeNull();

  });

  it('should re-parse custom amounts when SET_CONTRIBUTION_TYPE is fired for ANNUAL', () => {

    const contributionType = 'ANNUAL';
    const initialState = {
      contributionType: 'ONE_OFF',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: '1',
      isCustomAmount: true,
      error: null,
    };

    const changeContributionType = actions.setContributionType(contributionType, 'GBPCountries');
    const newState = reducer(initialState, changeContributionType);
    expect(newState.error).toEqual('TooLittle');

  });

  it('should handle SET_AMOUNT for one-off', () => {

    const amount = '42';
    const initialState = {
      contributionType: 'ONE_OFF',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: null,
      isCustomAmount: false,
      error: null,
    };

    const newState = reducer(initialState, actions.setAmount(amount));

    expect(newState.oneOffAmount).toEqual(amount);
    expect(newState.monthlyAmount).toEqual(initialState.monthlyAmount);
    expect(newState.annualAmount).toEqual(initialState.annualAmount);
    expect(newState.isCustomAmount).toEqual(false);
    expect(newState.error).toBeNull();

  });

  it('should handle SET_AMOUNT for monthly', () => {

    const amount = '42';
    const initialState = {
      contributionType: 'MONTHLY',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: null,
      isCustomAmount: false,
      error: null,
    };

    const newState = reducer(initialState, actions.setAmount(amount));

    expect(newState.monthlyAmount).toEqual(amount);
    expect(newState.oneOffAmount).toEqual(initialState.oneOffAmount);
    expect(newState.annualAmount).toEqual(initialState.annualAmount);
    expect(newState.isCustomAmount).toEqual(false);
    expect(newState.error).toBeNull();

  });

  it('should handle SET_AMOUNT for annual', () => {

    const amount = '42';
    const initialState = {
      contributionType: 'ANNUAL',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: null,
      isCustomAmount: false,
      error: null,
    };

    const newState = reducer(initialState, actions.setAmount(amount));

    expect(newState.annualAmount).toEqual(amount);
    expect(newState.oneOffAmount).toEqual(initialState.oneOffAmount);
    expect(newState.monthlyAmount).toEqual(initialState.monthlyAmount);
    expect(newState.isCustomAmount).toEqual(false);
    expect(newState.error).toBeNull();

  });

  it('should handle SET_AMOUNT_FOR_CONTRIBUTION_TYPE for one-off', () => {


    const amount = '42';
    const initialState = {
      contributionType: 'X',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: null,
      isCustomAmount: false,
      error: null,
    };

    const contributionType = 'ONE_OFF';

    const newState = reducer(initialState, actions.setAmountForContributionType(contributionType, amount));

    expect(newState.contributionType).toEqual(initialState.contributionType);
    expect(newState.oneOffAmount).toEqual(amount);
    expect(newState.monthlyAmount).toEqual(initialState.monthlyAmount);
    expect(newState.annualAmount).toEqual(initialState.annualAmount);
    expect(newState.isCustomAmount).toEqual(false);
    expect(newState.error).toBeNull();

  });

  it('should handle SET_AMOUNT_FOR_CONTRIBUTION_TYPE for monthly', () => {

    const amount = '42';
    const initialState = {
      contributionType: 'X',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: null,
      isCustomAmount: false,
      error: null,
    };

    const contributionType = 'MONTHLY';

    const newState = reducer(initialState, actions.setAmountForContributionType(contributionType, amount));

    expect(newState.contributionType).toEqual(initialState.contributionType);
    expect(newState.oneOffAmount).toEqual(initialState.oneOffAmount);
    expect(newState.monthlyAmount).toEqual(amount);
    expect(newState.annualAmount).toEqual(initialState.annualAmount);
    expect(newState.isCustomAmount).toEqual(false);
    expect(newState.error).toBeNull();

  });

  it('should handle SET_AMOUNT_FOR_CONTRIBUTION_TYPE for annual', () => {

    const amount = '42';
    const initialState = {
      contributionType: 'X',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: null,
      isCustomAmount: false,
      error: null,
    };

    const contributionType = 'ANNUAL';

    const newState = reducer(initialState, actions.setAmountForContributionType(contributionType, amount));

    expect(newState.contributionType).toEqual(initialState.contributionType);
    expect(newState.oneOffAmount).toEqual(initialState.oneOffAmount);
    expect(newState.monthlyAmount).toEqual(initialState.monthlyAmount);
    expect(newState.annualAmount).toEqual(amount);
    expect(newState.isCustomAmount).toEqual(false);
    expect(newState.error).toBeNull();

  });

  it('should handle SET_CUSTOM_AMOUNT', () => {

    const amount = '42';
    const initialState = {
      contributionType: 'ONE_OFF',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: null,
      isCustomAmount: false,
      error: null,
    };

    const newState = reducer(initialState, actions.setCustomAmount(amount, 'GBPCountries'));

    expect(newState.customAmount).toEqual(Number(amount));
    expect(newState.isCustomAmount).toEqual(true);
    expect(newState.error).toBeNull();

  });

  it('should handle SET_CUSTOM_AMOUNT errors', () => {

    const amount = 'notanumber';
    const initialState = {
      contributionType: 'ONE_OFF',
      oneOffAmount: '50',
      monthlyAmount: '5',
      annualAmount: '50',
      customAmount: null,
      isCustomAmount: false,
      error: null,
    };

    const newState = reducer(initialState, actions.setCustomAmount(amount, 'GBPCountries'));

    expect(newState.customAmount).toBeNull();
    expect(newState.error).toEqual('ParseError');

  });

});
