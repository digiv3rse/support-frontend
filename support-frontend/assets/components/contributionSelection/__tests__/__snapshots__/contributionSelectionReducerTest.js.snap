// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contributions Selection reducer should not change the state for an action with the wrong scope 1`] = `
Object {
  "annualAmount": "50",
  "contributionType": "MONTHLY",
  "customAmount": null,
  "error": null,
  "isCustomAmount": false,
  "monthlyAmount": "5",
  "oneOffAmount": "50",
}
`;

exports[`Contributions Selection reducer should not change the state for an action with the wrong scope 2`] = `
Object {
  "annualAmount": "50",
  "contributionType": "MONTHLY",
  "customAmount": null,
  "error": null,
  "isCustomAmount": false,
  "monthlyAmount": "5",
  "oneOffAmount": "50",
}
`;

exports[`Contributions Selection reducer should not change the state for an action with the wrong scope 3`] = `
Object {
  "annualAmount": "50",
  "contributionType": "MONTHLY",
  "customAmount": null,
  "error": null,
  "isCustomAmount": false,
  "monthlyAmount": "5",
  "oneOffAmount": "50",
}
`;

exports[`Contributions Selection reducer should return the initial state 1`] = `
Object {
  "annualAmount": "50",
  "contributionType": "MONTHLY",
  "customAmount": null,
  "error": null,
  "isCustomAmount": false,
  "monthlyAmount": "5",
  "oneOffAmount": "50",
}
`;

exports[`Contributions Selection reducer should return the initial state for another country group 1`] = `
Object {
  "annualAmount": "50",
  "contributionType": "MONTHLY",
  "customAmount": null,
  "error": null,
  "isCustomAmount": false,
  "monthlyAmount": "15",
  "oneOffAmount": "50",
}
`;
